docs/dev-sdk.md

```diff
@@ -1,59 +1,199 @@
+* Use `get` for non-mutable calls.
+* Use `put` with `body: "*"` most calls that need to send a message to the SDK server.
+
+#### Documentation
+
+* It is _imperative_ that the documentation is correct since it is used to automatically generate the documentation for https://libopenstorage.github.io .The documentation for these values in the proto files can be in Markdown format.
+
+* **Documenting Messages**
+    * Document each value of the message.
+    * Do not use Golang style. Do not repeat the name of the variable in _Golang Camel Format_ in the comment to document it since the variable could be in other styles in other languages. For example:
+
+```proto
+// Provides volume's exclusive bytes and its total usage. This cannot be
+// retrieved individually and is obtained as part node's usage for a given
+// node.
+message VolumeUsage {
+   // id for the volume/snapshot
+  string volume_id = 1;
+  // name of the volume/snapshot
+  string volume_name = 2;
+  // uuid of the pool that this volume belongs to
+  string pool_uuid = 3;
+  // size in bytes exclusively used by the volume/snapshot
+  uint64 exclusive_bytes = 4;
+  //  size in bytes by the volume/snapshot
+  uint64 total_bytes = 5;
+  // set to true if this volume is snapshot created by cloudbackups
+  bool local_cloud_snapshot = 6;
+}
+```
+
+#### Deprecation
+
+* **NOTE:** Most importantly is that these APIs _must_ be supported forever once released. 
+* They will almost never be deprecated since at some point we will have many versions of the clients. So please be clear and careful on the API you create.
+* If we need to change or update, you can always **add**.
+
+Here is the process if you would like to deprecate:
+
+1. According to [proto3 Language Guide](https://developers.google.com/protocol-buffers/docs/proto3) set the value in the message to deprecated and add a `(deprecated)` string to the comment as follows:
 
-### Version
-Please request what the version number of the SDK should be changed to if at all.
+```proto
+// (deprecated) Field documentation here
+int32 field = 6 [deprecated = true];
+```
+
+2. Comment in the SDK_CHANGELOG that the value is deprecated.
+3. Provide at least two releases before removing support for that value in the message. Make sure to document in the release notes of the product the deprecation.
+4. Once at least two releases have passed. Reserve the field number as shown in the [proto3 Language Guide](https://developers.google.com/protocol-buffers/docs/proto3#reserved):
+
+```proto
+message Foo {
+  reserved 6;
+}
+```
+
+It is essential that no values override the field number when updating or replacing. From the guide:
+
+> Note: If you update a message type by entirely removing a field, or commenting it out, future users can reuse the field number when making their own updates to the type. This can cause severe issues if they later load old versions of the same .proto, including data corruption, privacy bugs, and so on.
+
```

Comment:
With regards to field numbers, in a scenario where there is a oneof or sub message you could use the following pattern 

```
  string s3_storage_class = 7;
  // Start at field number 200 to allow for expansion
  oneof credential_type {
    // Credentials for AWS/S3
    SdkAwsCredentialRequest aws_credential = 200;
    // Credentials for Azure
    SdkAzureCredentialRequest azure_credential = 201;
    // Credentials for Google
    SdkGoogleCredentialRequest google_credential = 202;
  }
```

Reply:
Nice I will add that